### PR TITLE
Avoid parallel reroutes in DiskThresholdMonitor

### LINKED ISSUE #40174: MockDiskUsagesIT.testRerouteOccursOnDiskPassingHighWatermark fails in CI
The MockDiskUsagesIT#testRerouteOccursOnDiskPassingHighWatermark failed in CI since the node was expected to have 4 shards assigned to it but had 5. https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+matrix-java-periodic/ES_BUILD_JAVA=java11,ES_RUNTIME_JAVA=java11,nodes=immutable&&linux&&docker/81/console

Reproduce line (did not reproduce locally):

```
./gradlew :server:integTest \
  -Dtests.seed=2A58B98CB4ABA8E \
  -Dtests.class=org.elasticsearch.cluster.routing.allocation.decider.MockDiskUsagesIT \
  -Dtests.method="testRerouteOccursOnDiskPassingHighWatermark" \
  -Dtests.security.manager=true \
  -Dtests.locale=prg \
  -Dtests.timezone=Asia/Damascus \
  -Dcompiler.java=11 \
  -Druntime.java=11
```

Watermark triggered:

```
  1> [2019-03-18T18:45:27,176][WARN ][o.e.c.r.a.DiskThresholdMonitor] [node_t2] high disk watermark [90%] exceeded on [BY2R46gDQlCOF6x8VIUBvA][][/dev/null] free: 0b[0%], shards will be relocated away from this node
  1> [2019-03-18T18:45:27,178][INFO ][o.e.c.r.a.DiskThresholdMonitor] [node_t2] rerouting shards: [high disk watermark exceeded on one or more nodes]
```

Number of shards check that triggers the test failure:

```
   > Throwable #1: java.lang.AssertionError: node1 has 5 shards
   > Expected: <5>
   >      but: was <4>
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.cluster.routing.allocation.decider.MockDiskUsagesIT.lambda$testRerouteOccursOnDiskPassingHighWatermark$2(MockDiskUsagesIT.java:123)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:862)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:836)
   > 	at org.elasticsearch.cluster.routing.allocation.decider.MockDiskUsagesIT.testRerouteOccursOnDiskPassingHighWatermark(MockDiskUsagesIT.java:114)
   > 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   > 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   > 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   > 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   > 	at java.base/java.lang.Thread.run(Thread.java:834)
   > 	Suppressed: java.lang.AssertionError: node1 has 5 shards
```

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the `DiskThresholdMonitor` limits the frequency with which it submits
reroute tasks, but it might still submit these tasks faster than the master can
process them if, for instance, each reroute takes over 60 seconds. This causes
a problem since the reroute task runs with priority `IMMEDIATE` and is always
scheduled when there is a node over the high watermark, so this can starve any
other pending tasks on the master.

This change avoids further updates from the monitor while its last task(s) are
still in progress, and it measures the time of each update from the completion
time of the reroute task rather than its start time, to allow a larger window
for other tasks to run.

Fixes #40174

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java [hunk 1]
```diff
@@ -92,88 +99,112 @@ private void warnAboutDiskIfNeeded(DiskUsage usage) {
...
-                    nodeHasPassedWatermark.remove(node);
-                }
+
+        if (checkInProgress.compareAndSet(false, true) == false) {
+            logger.info("skipping monitor as a check is already in progress");
+            return;
+        }
+
+        final ImmutableOpenMap<String, DiskUsage> usages = info.getNodeLeastAvailableDiskUsages();
+        if (usages == null) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java]: Probably best to look at this bit ignoring whitespace changes - I removed a level of indentation.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java]: good tip

**Review by ywelsch — COMMENTED** [review]
I'm still not super happy that we are sending a task to be executed at priority IMMEDIATE. I would rather have this call `RoutingService`. In that case, we could also avoid this whole business of tracking whether there is already a call in progress (that's taken care of by RoutingService). WDYT?

**Comment by DaveCTurner** [comment]
I agree on the priority thing, but the `RoutingService` still uses `HIGH` priority and doesn't offer a notification on completion to keep the frequency low. I could add such a thing if you'd like?

**Comment by ywelsch** [comment]
> I agree on the priority thing, but the RoutingService still uses HIGH priority and doesn't offer a notification on completion to keep the frequency low. I could add such a thing if you'd like?

I think HIGH priority is ok for now. I wonder why we need the notification on completion. What does it keep the frequency low of? If we're batching calls, it's fine to have multiple pending attempts?

**Comment by DaveCTurner** [comment]
Ok I have added to the `RoutingService` the ability to listen for completion, and adjusted the `DiskThresholdMonitor` to make use of this. @ywelsch would you take another look?

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 2]
```diff
@@ -379,7 +379,7 @@ public void clusterStatePublished(ClusterChangedEvent clusterChangedEvent) {
                 if (logger.isTraceEnabled()) {
                     logger.trace("{}, scheduling a reroute", reason);
                 }
-                routingService.reroute(reason);
+                routingService.reroute(reason, ActionListener.wrap(() -> logger.trace("{}, reroute completed", reason)));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: this also logs the same line on an exception :/
I would prefer two different log lines, and the failure one with the exception (same for other places in this PR)

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java [hunk 3]
```diff
@@ -92,88 +104,112 @@ private void warnAboutDiskIfNeeded(DiskUsage usage) {
...
-            String explanation = "";
-
-            // Garbage collect nodes that have been removed from the cluster
-            // from the map that tracks watermark crossing
-            ObjectLookupContainer<String> nodes = usages.keys();
-            for (String node : nodeHasPassedWatermark) {
-                if (nodes.contains(node) == false) {
-                    nodeHasPassedWatermark.remove(node);
-                }
+
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java]: assert that rerouteAction is set?

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left two small asks. Looking good o.w.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java]: Fixed in c1d6ee0

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: image

Fixed in ce5946b

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/2

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**